file/src/main/scala/akka/stream/alpakka/file/impl/archive/TarReaderStage.scala

```diff
@@ -37,12 +36,15 @@ private[file] class TarReaderStage
           extends SubSourceOutlet[ByteString]("fileOut")
           with TarReaderStage.SourceWithTimeout
 
-      readHeader(ByteString.empty)
+      setHandlers(flowIn, flowOut, new CollectHeader(ByteString.empty))
 
       def readHeader(buffer: ByteString): Unit = {
         if (buffer.length >= TarArchiveEntry.headerLength) {
           readFile(buffer)
-        } else setHandlers(flowIn, flowOut, new CollectHeader(buffer))
+        } else {
+          if (!hasBeenPulled(flowIn)) pull(flowIn)
```

Comment:
For the problematic files/chunk size scenario we'd get here, without `flowIn` being pulled, and then switch to `CollectHeader` waiting to get the next chunk.